odoo/addons/base/models/res_users.py

```diff
@@ -199,6 +199,85 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
             return len(groups) if count else groups.ids
         return super(Groups, self)._search(args, offset=offset, limit=limit, order=order, count=count, access_rights_uid=access_rights_uid)
 
+    def _get_recursive_implied_groups(self, recursive_groups):
+        """ Returns the implied groups of the group being passed (including the
+            passed group itself)
+        """
+        groups = self.implied_ids.filtered(
+            lambda grp:
+            grp.category_id == self.category_id and
+            grp not in recursive_groups
+        )
+        if groups:
+            recursive_groups |= groups
+            recursive_groups = groups._get_recursive_implied_groups(recursive_groups)
+        return recursive_groups
+
+    def _prepare_warning_for_group_inheritance(self, user):
+        if not self:
+            return {}
+        inherited_groups = {}
+        self = self._origin
+        for selected_group in self:
+            # Prepare dictionary, with key = selected group, values = all the groups that matches their
+            # implied groups' category with selected group's category
+            inherited_groups[selected_group] = self.filtered(
+                lambda group:
+                selected_group.category_id in group.implied_ids.category_id and
+                selected_group.category_id != group.category_id and
+                selected_group not in group.implied_ids
+            )
+        warnings = []
+        for selected_group, other_groups in inherited_groups.items():
+            if other_groups:
```

Comment:
to remove indentation level, it's better to continue asap
```suggestion
            if not other_groups:
                continue
```